Feature dynamic hba

t0k4rt opened the pull request:
Hello, 
We want to deploy Stolon in our architecture, but we noticed that connection to su and repl users are open to the world which is (imho) not good for a secure cluster. 
We can deal with this with a firewall rule but I would prefer this to be setup correctly in the pg_hba.conf.
I saw a todo in your code concerning this issue so I decide to propose a fix. I'm not goloang developer nor sure it will match your coding standard or code organization wishes ... but it's a start.

The code simply read the cluster data generate a list of pghba entries and merges it with the `db.Spec.PGHBA`.

I found this to be the easiest solution.

Important to notice, it does not compile, I've got the current error: 
```
# github.com/sorintlab/stolon/vendor/github.com/sorintlab/tcpkeepalive
gopath/src/github.com/sorintlab/stolon/vendor/github.com/sorintlab/tcpkeepalive/tcpkeepalive.go:14:10: undefined: setIdle
gopath/src/github.com/sorintlab/stolon/vendor/github.com/sorintlab/tcpkeepalive/tcpkeepalive.go:22:10: undefined: setCount
gopath/src/github.com/sorintlab/stolon/vendor/github.com/sorintlab/tcpkeepalive/tcpkeepalive.go:29:10: undefined: setInterval
```

I also changed the default su and repl connection to localhost.

I hope it can be useful for you ! 
Don't hesitate to tell me how I can improve this.
Thanks for this product !
Alexandre

t0k4rt commented:
./test does not launch on OSX either

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1512,7 +1513,8 @@ func (p *PostgresKeeper) postgresKeeperSM(pctx context.Context) {
 		log.Infow("postgres hba entries changed, reloading postgres instance")
-		pgm.SetHba(db.Spec.PGHBA)
+		// Dynamicly generate hba auth from clusterData
+		pgm.SetHba(append(db.Spec.PGHBA, p.generateHBA(cd)...))
```
The generated hba entries should be inserted before the user provided hba entries.

sgotti replied:
don't append again here but do the append directly inside `generateHBA`.

t0k4rt commented:
No Problems, I'll change that !
I'm not experienced at all with go nor go vendors management, so please excuse this poor (yet) PR.

sgotti commented:
> I'm not experienced at all with go nor go vendors management, so please excuse this poor (yet) PR.

don't worry and welcome to the fantastic world of Go! Feel free to ask for any doubt (here on on gitter).

t0k4rt commented:
Hi, I updated the code as requested.
I added an optional switch defaultReplAccess that can take two options: "all" (default value) or "standby"

In case user select all:
* master will have its default hba with:
  * repl access to all ips (0.0.0.0/0)
  * su access to all ips (0.0.0.0/0)
* slave will have its default hba with:
  * su access to all ips (0.0.0.0/0)
  * no replication access since slave does not need it

In case user select standby:
* master will have its default hba with:
  * repl access to all standby ips (x.x.x.x/32)
  * su access to localhost (127.0.0.1/32)
  * su access to all standby ips (x.x.x.x/32)
* slave will have its default hba with:
  * su access to localhost (127.0.0.1/32) 
  * no replication access since slave does not need it

Maybe I should seperate repl and su options and create defaultReplAccess / defaultSuAccess ?

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -89,6 +89,7 @@ type config struct {
 	pgSUPasswordFile        string
 	pgInitialSUUsername     string
 	pgInitialSUPasswordFile string
+	defaultReplAccess       string
```
Sorry, I meant to set it as a cluster spec option, so it could be set globally: https://github.com/sorintlab/stolon/blob/master/pkg/cluster/cluster.go#L186

t0k4rt replied:
Not a problem I'll change that

t0k4rt replied:
Changes done

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1762,51 @@ func keeper(c *cobra.Command, args []string) {
+			fmt.Sprintf("host all %s %s %s\n", p.pgSUUsername, "::0/0", p.pgSUAuthMethod),
+		)
+		// allow only for master server replication access
+		if role == common.RoleMaster {
```
This is a change from the current behaviour and it's breaking the standby cluster feature and related tests. So it should be removed.

t0k4rt replied:
I don't understand why it breaks. 
Master pg_hba should only allow replication user to connect from standby servers.
Disbaled or enabling replication user on standby server should not have any effects. 
When a standby server is promoted, its pg_hba should be automatically updated and would contain the right rules.

sgotti replied:
Stolon can also be in standby cluster mode. In this case the master keeper manages a standby instance that will sync with an external instance. So the other keepers will sync from it (cascading replication) and the above check will break.

I suggested to just remove it because to keep backward compatibility with "all" mode we have to keep the previous rules.

t0k4rt replied:
Ok, changes done. It should be ok !

t0k4rt commented:
I got during tests this issue, So I'm going to do more tests
```
utils.go:121: [keeper c594fd73]: panic: runtime error: invalid memory address or nil pointer dereference
	utils.go:121: [keeper c594fd73]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xadb63f]
```

sgotti commented on pkg/cluster/cluster.go:
```diff
@@ -183,6 +184,21 @@ type StandbySettings struct {
+	// Allow hba access to all members (0.0.0.0/0)
+	PgHBAAccessAll PgHBAAccessMode = "all"
+	// Allow hba access to standby server only
+	PgHBAAccessStandby PgHBAAccessMode = "standby"
```
Stale field?

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1756,45 @@ func keeper(c *cobra.Command, args []string) {
+	// Dynamically generate hba auth from clusterData
+	var hbaBase []string
+
+	if cluster.PgHBAAccessAll == *cd.Cluster.Spec.PgHBAAccess {
```
Use cd.Cluster.DefSpec().PgHBAAccess or you'll miss the default values when not provided

t0k4rt replied:
Done !

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1756,45 @@ func keeper(c *cobra.Command, args []string) {
+	// Dynamically generate hba auth from clusterData
+	var hbaBase []string
+
+	if cluster.PgHBAAccessAll == *cd.Cluster.Spec.PgHBAAccess {
```
use a switch statement here?

t0k4rt replied:
Done !

sgotti commented on pkg/cluster/cluster.go:
```diff
@@ -243,6 +259,10 @@ type ClusterSpec struct {
 	// Standby setting when role is standby
 	StandbySettings *StandbySettings `json:"standbySettings,omitempty"`
+	// Set default hba auth for replication user and superuser.
+	// Values can be "all" or "standby", "all" allow access to all ips, "standby" restrict master access to standby servers ips.
```
standby or strict?

t0k4rt replied:
IMHO it's better strict

sgotti commented on pkg/cluster/cluster.go:
```diff
@@ -243,6 +259,10 @@ type ClusterSpec struct {
+	// Set default hba auth for replication user and superuser.
+	// Values can be "all" or "standby", "all" allow access to all ips, "standby" restrict master access to standby servers ips.
+	// default is "all"
+	PgHBAAccess *PgHBAAccessMode `json:"pgHBAAccess,omitempty"`
```
I'll call this `DefaultSUReplAccessMode`, longer but a bit clearer. Other Suggestions?

t0k4rt replied:
We could use:
* DefaultSUReplAccessMode as option
* SUReplAccessMode as type
* DefaultSUReplAccess as default value
* SUReplAccessAll and SUReplAccessStrict as option names 
Is it ok for you ?

sgotti replied:
Seems Good.

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1510,9 +1511,11 @@ func (p *PostgresKeeper) postgresKeeperSM(pctx context.Context) {
 
-	if !reflect.DeepEqual(db.Spec.PGHBA, pgm.CurHba()) {
+	// Dynamicly generate hba auth from clusterData
+	new_hba := p.generateHBA(cd, localRole, db.Spec.PGHBA)
```
s/new_hba/newHBA/

t0k4rt commented:
I've got an issue when I try to update my ClusterSpec, 
the key "defaultSUReplAccessMode": "strict" is properly set but not persisted. 
Any ideas ?

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,45 @@ func keeper(c *cobra.Command, args []string) {
+		// allow for all servers su access
+		hbaBase = append(
+			hbaBase,
+			fmt.Sprintf("host all %s %s %s\n", p.pgSUUsername, "127.0.0.1/32", p.pgReplAuthMethod),
```
This is only from 127.0.0.1/32 not from all the standbys or am I missing something?

sgotti commented on pkg/postgresql/postgresql.go:
```diff
@@ -713,21 +713,10 @@ func (p *Manager) writePgHba() error {
-
 	if p.hba != nil {
 		for _, e := range p.hba {
 			f.WriteString(e + "\n")
```
You have forgot to handle the "else" case below inside generateHBA. Here it'll always be skipped since p.hba is never empty.

t0k4rt replied:
Ok, this means that previously, when p.hba was empty, the generated hba was:

```
host replication repluser 0.0.0.0/0 md5
host replication repluser ::0/0 md5
host all superuser 0.0.0.0/0 md5
host all superuser ::0/0 md5
host all all 0.0.0.0/0 md5
host all all ::0/0 md5
```
which is same as this:
```
host all all 0.0.0.0/0 md5
host all all ::0/0 md5
```
but simpler so I think this should be the result in "all" access mode.

In "strict" access mode we should not use this rule.

sgotti replied:
Not always, the first lines can have a different auth method than md5 (currently only trust but in future other methods could be added), while the last lines are added only if there's no user provided pgHBA and always have md5. I'd prefer to not change the current logic.

t0k4rt replied:
Yes you're right ! 

For the strict mode, I still would prefer not to add this rule.

sgotti replied:
It may be confusing but these are two separate things. The first lines (the one affected by this patch to set the DefaultSuReplAccessMode) are for instance replication, the latter is a default when the user doesn't provide a PGHBA. So I'll leave them distinct and not make DefaultSuReplAccessMode change this.

t0k4rt replied:
Ok I understand !
Can I plan a DefaultUserAccessMode feature ?

sgotti replied:
@t0k4rt you can just do the same proving an empty (not nil) pgHBA in the clusterspec

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,45 @@ func keeper(c *cobra.Command, args []string) {
+		hbaBase = append(
+			hbaBase,
+			fmt.Sprintf("host all %s %s %s\n", p.pgSUUsername, "0.0.0.0/0", p.pgSUAuthMethod),
+			fmt.Sprintf("host all %s %s %s\n", p.pgSUUsername, "::0/0", p.pgSUAuthMethod),
```
remove `\n` in all the lines or an empty line will be added

t0k4rt replied:
Done

sgotti commented:
> the key "defaultSUReplAccessMode": "strict" is properly set but not persisted.

What do you mean by "not persisted"? I tried a stolonctl patch and it has been applied to the cluster spec and the pghba file has been updated (with some problems, see my comments) by the keepers.

t0k4rt commented:
@sgotti Ok the not persisted issue is because I've got a cluster running two different versions of the keeper. The other keeper does not know the new key so it might removes it.

sgotti commented:
> @sgotti Ok the not persisted issue is because I've got a cluster running two different versions of the keeper. The other keeper does not know the new key so it might removes it.

The unique component that writes the cluster data is the sentinel (and stolonctl) so perhaps you had an old sentinel.

t0k4rt commented on pkg/postgresql/postgresql.go:
```diff
@@ -713,30 +713,14 @@ func (p *Manager) writePgHba() error {
-	// Matched local replicaton connections are for replUsername user with md5 auth
+	// Matched local replication connections are for replUsername user with md5 auth
 	f.WriteString(fmt.Sprintf("local postgres %s %s\n", p.suUsername, p.suAuthMethod))
 	f.WriteString(fmt.Sprintf("local replication %s %s\n", p.replUsername, p.replAuthMethod))
```
Done

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,63 @@ func keeper(c *cobra.Command, args []string) {
+			"host all all 0.0.0.0/0 md5",
+			"host all all ::0/0 md5",
+		)
+	}
```
this must be at the end. After the stolon generated rules.

t0k4rt replied:
Done

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,63 @@ func keeper(c *cobra.Command, args []string) {
+	}
+
+	// when no custom hba is set, we default to open to all
+	if len(userHba) == 0 {
```
don't use len but check if it's nil or not like currently done. So you can distinguis between a not provided (nil) or a provided but empty pgHBA.

t0k4rt replied:
Done

sgotti commented on pkg/cluster/cluster.go:
```diff
@@ -243,6 +257,10 @@ type ClusterSpec struct {
 	// Standby setting when role is standby
 	StandbySettings *StandbySettings `json:"standbySettings,omitempty"`
+	// Set default hba auth for replication user and superuser.
+	// Values can be "all" or "strict", "all" allow access to all ips, "strict" restrict master access to standby servers ips.
```
s/to/from/g

sgotti commented on pkg/cluster/cluster.go:
```diff
@@ -243,6 +257,10 @@ type ClusterSpec struct {
 	ExistingConfig *ExistingConfig `json:"existingConfig,omitempty"`
 	// Standby setting when role is standby
 	StandbySettings *StandbySettings `json:"standbySettings,omitempty"`
+	// Set default hba auth for replication user and superuser.
```
`Define the mode of the default hba rules needed for replication by standby keepers (the su and repl auth methods will be the one provided in the keeper command line options)`

t0k4rt replied:
Done

t0k4rt commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,65 @@ func keeper(c *cobra.Command, args []string) {
+		)
+		// allow only for master replication access
+		if role == common.RoleMaster {
+			// we should not allow replication acces for 127.0.0.1 since it can be then forwarded through proxy.
```
No problems, I'll check that !
I'll update the cluster spec too
Concerning an integration test I'll check too but not sure I'll managed to do something good enough ...

sgotti replied:
> Concerning an integration test I'll check too but not sure I'll managed to do something good enough ...

Don't worry, I could add something later. Unfortunately integration tests of distributed system is quite tedious and complex...

t0k4rt commented:
@sgotti Hello, everything should be fixed.
Have you any other feedback ? 
thanks !

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,94 @@ func keeper(c *cobra.Command, args []string) {
+	if currentDb == nil {
+		panic(fmt.Sprintf("Unable to get current keeper local DB with UID %s", p.keeperLocalState.UID))
+	}
+
```
* you already have the `db` in the caller just pass it to generateHBA without all of the above (and avoid calls to panic)
* `role common.Role` isn't used anymore
* rename `userHba` to `specHBA` to be clearer.

generateHBA should become:
```
func (p *PostgresKeeper) generateHBA(cd *cluster.ClusterData, db *cluster.DB, specHBA []string) []string {
```

t0k4rt replied:
Ok but since I get specHBA from `db.spec.PGHBA` I could simplify and just only call `func (p *PostgresKeeper) generateHBA(cd *cluster.ClusterData, db *cluster.DB) []string` no ?

sgotti replied:
right!

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,94 @@ func keeper(c *cobra.Command, args []string) {
+		}
+	}
+
+	// when no custom hba is set, we default to open to all
```
let's keep the original comment, this isn't completely correct.

```
// By default, if no custom pg_hba entries are provided, accept
// connections for all databases and users with md5 auth
```

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,86 @@ func keeper(c *cobra.Command, args []string) {
+		// we should not allow replication acces for 127.0.0.1 since it can be then forwarded through proxy.
+		computedHBA = append(
+			computedHBA,
+			fmt.Sprintf("host all %s %s %s", p.pgSUUsername, "127.0.0.1/32", p.pgSUAuthMethod),
```
shouldn't this be `127.0.0.0/8` ?

t0k4rt replied:
yes that's better !

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,86 @@ func keeper(c *cobra.Command, args []string) {
+	computedHBA := []string{
+		fmt.Sprintf("local postgres %s %s\n", p.pgSUUsername, p.pgSUAuthMethod),
+		fmt.Sprintf("local replication %s %s\n", p.pgReplUsername, p.pgReplAuthMethod),
+	}
```
remove `\n`

t0k4rt replied:
Done

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,86 @@ func keeper(c *cobra.Command, args []string) {
+	// Case strict access mode
+	case cluster.SUReplAccessStrict:
+		// allow for all servers su access
+		// we should not allow replication acces for 127.0.0.1 since it can be then forwarded through proxy.
```
Sorry but I forgot why this is needed. The comment isn't really clear and doesn't looks related to the added rules.

t0k4rt replied:
True,  it's changed

sgotti commented on cmd/keeper/keeper.go:
```diff
@@ -1754,3 +1757,83 @@ func keeper(c *cobra.Command, args []string) {
+		computedHBA = append(
+			computedHBA,
+			fmt.Sprintf("host all %s %s %s", p.pgSUUsername, "127.0.0.0/8", p.pgSUAuthMethod),
+			fmt.Sprintf("host all %s %s %s", p.pgSUUsername, "::1/128", p.pgSUAuthMethod),
```
Sorry but I don't remember why these line are needed? If really needed a comment explaining the reason will be useful.

t0k4rt replied:
Actually It's not mandatory since we already have
`fmt.Sprintf("local postgres %s %s", p.pgSUUsername, p.pgSUAuthMethod)`
I can drop this

t0k4rt replied:
Done

sgotti commented:
@t0k4rt Can you import the unit test I made in https://github.com/sgotti/stolon/commit/cfc5a03709793f0a76ba5fa5aa95d339a6286330 and squash and rebase against the current master? Thanks!

t0k4rt commented:
@sgotti Hello, I couldn't manage to cherry pick your commit so I included the test manually. I hope it's okay.

sgotti commented:
@t0k4rt unfortunately you have to rebase you branch on the current master and it'll be better if you squash all your commits in only one commit. If you have some problems doing it let me know.

t0k4rt commented:
@sgotti I think this should be ok, could you check ?

sgotti commented:
@t0k4rt can you please squash you commits in an unique one?

t0k4rt commented:
@sgotti I'm sorry, I merged my branch with your master to resolve conflict and squashed commits with rebase. But there are still conflict. Could you help me ?

sgotti commented:
@t0k4rt I rebased your branch using `git --rebase origin/master`, fixed two conflicts and did some minor changes to the doc, some comments and the commit title.

Take it here: https://github.com/sgotti/stolon/commit/1f4d4296394ab3b5863873c3ac3f388f4b8aa120

sgotti commented:
@t0k4rt sorry but it's not just my commit (I see 1157 changed files) so it's not rebased on current origin/master.

You should do something like this:

on your local branch:

`git fetch`
`git reset --hard origin/master`

then cherry pick my commit. The if you run `git rebase origin/master` you shouldn't see any error and change.

Then force push your branch to github to update the PR.

t0k4rt commented:
@sgotti is it okay now ? I started again from sorintlab/stolon master branch.

sgotti commented:
@t0k4rt We're almost there. 

I just noticed that the new functions in keeper.go are at the end of the file. I'll prefer to move them before but it's not a big deal.

I also see a difference in two comments:

```
--- a/pkg/cluster/cluster.go
+++ b/pkg/cluster/cluster.go
@@ -187,9 +187,9 @@ type StandbySettings struct {
 type SUReplAccessMode string
 
 const (
-       // Allow access from every host
+       // Allow hba access to all members (0.0.0.0/0)
        SUReplAccessAll SUReplAccessMode = "all"
-       // Allow access from standby server IPs only
+       // Allow hba access to standby server only (strict mode)
        SUReplAccessStrict SUReplAccessMode = "strict"
 )
```

t0k4rt commented:
@sgotti I put all *PostgresKeeper functions with the others and put the right comments. This should look good !

sgotti commented:
@t0k4rt Great! Thanks a lot of work and patience! Merging.

t0k4rt commented:
@sgotti Grazie mille !
